### PR TITLE
Add ascend and descend while flying

### PR DESCRIPTION
**Pull request by Lurk390**
Fixes issue #82 

Added the ability to ascend and descend while flying. Pressing the spacebar will ascend the player, while pressing left shift will descend them. Added instructions to `README.md` as well.

### CODE REVIEW

File: tempus_fugit_minecraft/main.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from pyglet.gl import *
 from tempus_fugit_minecraft.window import Window, WINDOW_WIDTH, WINDOW_HEIGHT
 
+
```
Comment (criticalglitch) [inline tempus_fugit_minecraft/main.py:7]: Why is this extra line of blank space necessary? One blank line between the imports and the first function should be sufficient right?

File: tests/test_cloud.py
Comment (criticalglitch) [inline tests/test_cloud.py:1]: Please delete this file. `test_model.py` is the former `test_cloud.py` as of 5894f9d.

File: tempus_fugit_minecraft/model.py [hunk 2]
```diff
@@ -55,14 +55,14 @@ class Model(object):
...
-        
+
         # A Batch is a collection of vertex lists for batched rendering.
         self.batch = Batch()
 
         # A TextureGroup manages an OpenGL texture.
         self.group = TextureGroup(image.load(TEXTURE_PATH).get_texture())
 
-        # A mapping from position to the block at that position.
+        # A mapping from position to the texture of the block at that position.
```
Comment (criticalglitch) [inline tempus_fugit_minecraft/model.py:65]: Please undo this change, as of 5894f9d, the `self.world` property is a map of position tuples to Block objects, not to _just_ their textures.

**Review by criticalglitch — CHANGES_REQUESTED** [review]
First, please make sure you've rebased on top of master.

Second, while testing this I found a major bug where the player could fly through the bottom of the world because we weren't properly handling the collision logic.

Third, I have some questions and suggestions below.

File: tempus_fugit_minecraft/player.py [hunk 3]
```diff
@@ -149,7 +152,7 @@ def adjust_sight(self, dx, dy):
 
     def current_speed(self):
         return self.FLYING_SPEED if self.flying else self.walking_speed
-    
+
     def toggle_flight(self):
```
Comment (xieWright) [inline tempus_fugit_minecraft/player.py:156]: Probably best to link the issue number and leave an embedded comment on the function.

**Review by xieWright — COMMENTED** [review]
When flying, are we supposed to be able to go through the ground? Just checked that. Else seems like things are working.

**Comment by criticalglitch** [comment]
> When flying, are we supposed to be able to go through the ground? Just checked that. Else seems like things are working.

This is the issue [here](https://github.com/WSUCEG-7140/Tempus_Fugit_Minecraft/pull/83#discussion_r1266062186), the fix is to remove that block of code the comment is attached to.

File: tempus_fugit_minecraft/main.py (hunk 1 above)
Comment (Lurk390) [reply tempus_fugit_minecraft/main.py:7]: According to PEP8 formatting style, functions should have 2 newlines before them if they are not in a class. See this link: https://peps.python.org/pep-0008/#blank-lines

File: tests/test_cloud.py
Comment (Lurk390) [reply tests/test_cloud.py:1]: I think I made a mistake when rebasing, hence the outdated code scattered about. Will fix.

File: tempus_fugit_minecraft/window.py [hunk 4]
```diff
@@ -211,16 +211,23 @@ def on_key_press(self, symbol: int, modifiers: int) -> None:
...
+        elif symbol == key.LSHIFT:
+            if self.model.player.flying:
+                self.model.player.descend = True
+
         if symbol == key.SPACE:
-            self.model.handle_jump()
+            if self.model.player.flying:
+                self.model.player.ascend = True
+            else:
+                self.model.handle_jump()
```
[the remark this answers is not in the document]
Comment (Lurk390) [reply tempus_fugit_minecraft/window.py:230]: Good idea, will implement this.

File: tempus_fugit_minecraft/model.py (hunk 2 above)
Comment (Lurk390) [reply tempus_fugit_minecraft/model.py:65]: Will do.

File: tempus_fugit_minecraft/model.py [hunk 5]
```diff
@@ -475,9 +474,17 @@ def update(self, dt: float) -> None:
...
             self.sector = sector
-        
+
+        if self.player.flying:
+            if self.player.ascend:
+                self.player.position = self.player.position[0], self.player.position[1] + dt * self.player.FLYING_SPEED, \
+                    self.player.position[2]
+            if self.player.descend:
+                self.player.position = self.player.position[0], self.player.position[1] - dt * \
+                                                                self.player.FLYING_SPEED, self.player.position[2]
```
[the remark this answers is not in the document]
Comment (Lurk390) [reply tempus_fugit_minecraft/model.py:484]: Yep, looks like I forgot to remove it when I was doing merge conflicts.

File: tests/test_flying.py
Comment (criticalglitch) [inline tests/test_flying.py:1]: Suggestion, merge this, test_speed.py and test_pause_menu.py into a generic test_window.py class with a single TestWindow class that has all of these tests.

In all three we're specifically testing the interactions of the window class, so we should have them all in a common TestWindow class.

File: tests/test_model.py [hunk 6]
```diff
@@ -8,11 +8,13 @@
 from tempus_fugit_minecraft.utilities import WORLD_SIZE
 import random
 
+
 @pytest.fixture(scope="class")
 def model():
-    pyglet.options['audio'] = ('silent')
+    pyglet.options['audio'] = 'silent'
```
Comment (criticalglitch) [inline tests/test_model.py:14]: Please undo this change.  From what I'm reading in the [pyglet options](https://pyglet.readthedocs.io/en/pyglet-1.5-maintenance/programming_guide/options.html) docs, I think we need this to be a tuple of just the one item since it says it should be a _sequence_.

**Review by criticalglitch — CHANGES_REQUESTED** [review]
Only a few issues left from me, only in test code thankfully.

File: tests/test_model.py (hunk 6 above)
Comment (Lurk390) [reply tests/test_model.py:14]: Oh okay, I thought the parenthesis were redundant so I removed them. Will revert.

**Review by xieWright — APPROVED** [review]
Looks like it's working. Not going through the floor anymore and passed all the tests.

**Review by w1158mxs — APPROVED** [review]
Good job, Thank you!
